### PR TITLE
ENT-4790: Limit compliance recalculations

### PR DESCRIPTION
**Pull request by Januson**

### CODE REVIEW

**Review by Ceiu — APPROVED** [review]
Code changes seem fine, though I would like someone who went over the design to give this a once-over as well.

File: src/main/java/org/candlepin/resource/ConsumerResource.java [hunk 1]
```diff
@@ -1589,24 +1591,36 @@ else if (toUpdate.getContentAccessMode() != null) {
...
+        if (complianceChangesMade) {
+            // this should update compliance on toUpdate, but not call the curator
+            complianceRules.getStatus(toUpdate, null, false, false);
         }
 
         return changesMade;
     }
 
+    private boolean complianceFactsChanged(ConsumerDTO updated, Consumer toUpdate) {
+        return !Objects.equals(ComplianceFacts.filter(updated.getFacts()), ComplianceFacts.of(toUpdate));
```
Comment (nikosmoum) [inline src/main/java/org/candlepin/resource/ConsumerResource.java:1621]: It's a bit unclear reading this that the `filter` and `of` methods are basically doing the same thing.
We could make the interface of that class a bit better by adding an overloaded `of(ConsumerDTO)` method, and make the `filter` method private.

File: src/main/java/org/candlepin/policy/js/compliance/hash/ComplianceFacts.java [hunk 2]
```diff
@@ -58,10 +58,20 @@ private String getFactKey() {
      * @return facts relevant for compliance
      */
     public static Collection<Map.Entry<String, String>> of(Consumer target) {
-        if (target.getFacts() == null) {
+        return filter(target.getFacts());
```
Comment (nikosmoum) [inline src/main/java/org/candlepin/policy/js/compliance/hash/ComplianceFacts.java:61]: let's check if target is null first

**Review by nikosmoum — CHANGES_REQUESTED** [review]
2 minor changes, but other than that this is great

**Comment by nikosmoum** [comment]
also there are some non-flaky test failures though that seem affected by the changes

**Review by nikosmoum — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/resource/ConsumerResource.java [hunk 3]
```diff
@@ -1589,24 +1591,36 @@ else if (toUpdate.getContentAccessMode() != null) {
...
+
+        if (sysPurposeChangesMade) {
             // this should update compliance on toUpdate, but not call the curator
-            complianceRules.getStatus(toUpdate, null, false, false);
             systemPurposeComplianceRules.getStatus(toUpdate, toUpdate.getEntitlements(), null, false);
+        }
 
-            Event event = eventBuilder.setEventData(toUpdate).buildEvent();
-            sink.queueEvent(event);
+        if (complianceChangesMade) {
```
[the remark this answers is not in the document]
Comment (Januson) [reply src/main/java/org/candlepin/resource/ConsumerResource.java:1613]: Updates without any compliance/syspurpose changes are ~20% faster.

File: src/main/java/org/candlepin/resource/ConsumerResource.java [hunk 4]
```diff
@@ -1503,13 +1504,14 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
...
             .setEventData(toUpdate);
 
-
         // version changed on non-checked in consumer, or list of capabilities
         // changed on checked in consumer
         boolean changesMade = updateCapabilities(toUpdate, updated);
 
+        boolean complianceFactsChanged = complianceFactsChanged(updated, toUpdate);
+        boolean installedProductsChanged = checkForInstalledProductsUpdate(toUpdate, updated);
+        boolean complianceChangesMade = complianceFactsChanged || installedProductsChanged;
```
[the remark this answers is not in the document]
Comment (Januson) [reply src/main/java/org/candlepin/resource/ConsumerResource.java]: Done.

**Pull request merged**